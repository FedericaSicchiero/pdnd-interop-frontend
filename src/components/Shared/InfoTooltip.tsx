import React from 'react'
import { StyledTooltip } from './StyledTooltip'

type InfoTooltipProps = {
  label: string
}

export function InfoTooltip({ label }: InfoTooltipProps) {
  return (
    <StyledTooltip title={label}>
      <i className={`text-primary fs-6 bi bi-info-circle ms-2`} />
    </StyledTooltip>
  )
}
