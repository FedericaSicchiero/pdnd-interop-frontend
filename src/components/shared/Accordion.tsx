import React from 'react'
import {
  Accordion as MUIAccordion,
  AccordionDetails,
  AccordionSummary,
  Typography,
} from '@mui/material'
import ExpandMoreIcon from '@mui/icons-material/ExpandMore'

export type AccordionEntry = {
  summary: string | JSX.Element
  summarySecondary?: string | JSX.Element
  details: string | JSX.Element
}

type StyledAccordionProps = {
  entries: Array<AccordionEntry>
}

export function Accordion({ entries }: StyledAccordionProps) {
  return (
    <>
      {entries.map(({ summary, summarySecondary, details }, i: number) => (
        <MUIAccordion key={i}>
          <AccordionSummary
            expandIcon={<ExpandMoreIcon />}
            aria-controls={`panel-content-${i}`}
            id={`panel-header-${i}`}
            sx={{ px: 0 }}
          >
            <Typography
              component={'span'}
              sx={{
                flexShrink: summarySecondary ? 0 : 1,
                width: summarySecondary ? '40%' : 'auto',
              }}
            >
              {summary}
            </Typography>
            {summarySecondary && (
              <Typography component="span" color="text.secondary" variant="body2">
                {summarySecondary}
              </Typography>
            )}
          </AccordionSummary>
          <AccordionDetails>{details}</AccordionDetails>
        </MUIAccordion>
      ))}
    </>
  )
}
