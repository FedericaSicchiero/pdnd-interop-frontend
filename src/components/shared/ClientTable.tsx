import { ClientQueries } from '@/api/client'
import useGetClientActions from '@/hooks/useGetClientActions'
import { useJwt } from '@/hooks/useJwt'
import { useNavigateRouter } from '@/router'
import { Client, ClientKind } from '@/types/client.types'
import { Box, Button, Skeleton, Stack } from '@mui/material'
import React from 'react'
import { useTranslation } from 'react-i18next'
import ActionMenu from './ActionMenu'
import { Table, TableRow } from './Table'

type ClientTableRow = {
  client: Client
  clientKind: ClientKind
}

const ClientTableRow: React.FC<ClientTableRow> = ({ client, clientKind }) => {
  const { t } = useTranslation('common', { keyPrefix: 'actions' })
  const { navigate } = useNavigateRouter()
  const prefetch = ClientQueries.usePrefetchSingle()

  const { actions } = useGetClientActions(client)

  const handlePrefetch = () => {
    prefetch(client.id)
  }

  const handleInspect = () => {
    const path =
      clientKind === 'CONSUMER' ? 'SUBSCRIBE_CLIENT_EDIT' : 'SUBSCRIBE_INTEROP_M2M_CLIENT_EDIT'

    navigate(path, { params: { clientId: client.id } })
  }

  return (
    <TableRow cellData={[{ label: client.name }]}>
      <Button
        onPointerEnter={handlePrefetch}
        onFocusVisible={handlePrefetch}
        variant="outlined"
        size="small"
        onClick={handleInspect}
      >
        {t('inspect')}
      </Button>

      <Box component="span" sx={{ ml: 2, display: 'inline-block' }}>
        <ActionMenu actions={actions} />
      </Box>
    </TableRow>
  )
}

const ClientTableRowSkeleton: React.FC = () => {
  return (
    <TableRow cellData={[{ label: <Skeleton width={440} /> }]}>
      <Stack direction="row" sx={{ display: 'inline-flex' }}>
        <Skeleton sx={{ borderRadius: 1 }} variant="rectangular" width={100} height={35} />
        <Box
          sx={{
            ml: 4,
            mr: 2,
            display: 'inline-flex',
            alignItems: 'center',
            justifyContent: 'center',
          }}
        >
          <Skeleton variant="rectangular" width={4} />
        </Box>
      </Stack>
    </TableRow>
  )
}

interface ClientTableProps {
  clientKind: ClientKind
}

const ClientTable: React.FC<ClientTableProps> = ({ clientKind }) => {
  const { t } = useTranslation('client')
  const { t: tCommon } = useTranslation('common', { keyPrefix: 'table.headData' })
  const { jwt } = useJwt()
  const { data: clients } = ClientQueries.useGetAll({
    kind: clientKind,
    consumerId: jwt?.organizationId,
  })

  const headLabels = [tCommon('clientName'), '']
  const isEmpty = clients && clients.length === 0

  return (
    <Table headLabels={headLabels} isEmpty={isEmpty} noDataLabel={t('noMultiDataLabel')}>
      {!isEmpty &&
        clients?.map((client) => (
          <ClientTableRow key={client.id} client={client} clientKind={clientKind} />
        ))}
    </Table>
  )
}

export const ClientTableSkeleton: React.FC = () => {
  const { t } = useTranslation('common', { keyPrefix: 'table.headData' })
  const headLabels = [t('clientName'), '']
  return (
    <Table headLabels={headLabels}>
      <ClientTableRowSkeleton />
      <ClientTableRowSkeleton />
      <ClientTableRowSkeleton />
      <ClientTableRowSkeleton />
      <ClientTableRowSkeleton />
    </Table>
  )
}

export default ClientTable
