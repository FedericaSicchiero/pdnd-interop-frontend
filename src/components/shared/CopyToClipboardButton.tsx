import React, { useState, useRef, useEffect } from 'react'
import { IconButton, Tooltip } from '@mui/material'
import type { IconButtonProps } from '@mui/material'
import ContentCopyIcon from '@mui/icons-material/ContentCopy'
import CheckIcon from '@mui/icons-material/Check'

interface CopyToClipboardProps extends Omit<IconButtonProps, 'onClick' | 'value'> {
  /** callback used to retrieve the text to be copied */
  value: (() => string) | string
  tooltipTitle?: string
}

const CopyToClipboard: React.FC<CopyToClipboardProps> = ({ value, tooltipTitle, ...props }) => {
  const [copied, setCopied] = useState(false)
  const copiedTimeoutRef = useRef<NodeJS.Timeout>()

  const handleCopyToClipboard = async () => {
    const valueToCopy = value instanceof Function ? value() : value

    if (tooltipTitle) {
      clearTimeout(copiedTimeoutRef.current)
      setCopied(true)
      copiedTimeoutRef.current = setTimeout(() => setCopied(false), 2000)
    }

    try {
      await navigator.clipboard.writeText(valueToCopy)
    } catch (err) {
      console.error(err)
    }
  }

  /** Clears the timeout on component unmount */
  useEffect(() => {
    return () => {
      clearTimeout(copiedTimeoutRef.current)
    }
  }, [])

  return (
    <Tooltip open={copied} arrow={true} title={tooltipTitle} placement="top">
      <IconButton color="primary" onClick={handleCopyToClipboard} {...props}>
        {copied && <CheckIcon fontSize="small" />}
        {!copied && <ContentCopyIcon fontSize="small" />}
      </IconButton>
    </Tooltip>
  )
}

export default CopyToClipboard
