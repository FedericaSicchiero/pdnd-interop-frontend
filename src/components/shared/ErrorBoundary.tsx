import React from 'react'
import { QueryErrorResetBoundary } from '@tanstack/react-query'
import { ErrorBoundary as _ErrorBoundary, FallbackProps } from 'react-error-boundary'

interface ErrorBoundaryProps {
  children: React.ReactNode
  FallbackComponent: React.ComponentType<FallbackProps>
}

export const ErrorBoundary: React.FC<ErrorBoundaryProps> = ({ children, FallbackComponent }) => {
  return (
    <QueryErrorResetBoundary>
      {({ reset }) => (
        <_ErrorBoundary onReset={reset} FallbackComponent={FallbackComponent}>
          {children}
        </_ErrorBoundary>
      )}
    </QueryErrorResetBoundary>
  )
}
