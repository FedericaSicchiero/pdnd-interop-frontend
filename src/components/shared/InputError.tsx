import React from 'react'
import { Box, Typography } from '@mui/material'

type InputErrorProps = {
  error: unknown
}

type Error = {
  message: string
}

export function InputError({ error }: InputErrorProps) {
  return (
    <Box>
      <Typography variant="caption" color="error">
        {(error as Error).message}
      </Typography>
    </Box>
  )
}
