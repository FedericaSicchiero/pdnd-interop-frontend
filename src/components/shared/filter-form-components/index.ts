export * from './FilterWrapper'
export * from './FilterTextField'
export * from './FilterAutocompleteSingle'
export * from './FilterAutocompleteMultiple'
