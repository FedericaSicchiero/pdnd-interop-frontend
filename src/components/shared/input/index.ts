export { Autocomplete } from './Autocomplete'
export { Select } from './Select'
