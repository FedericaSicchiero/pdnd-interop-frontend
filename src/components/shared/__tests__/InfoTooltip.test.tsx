import React from 'react'
import { render } from '@testing-library/react'
import { InfoTooltip } from '@/components/shared/InfoTooltip'

describe("Checks that InfoTooltip snapshots don't change", () => {
  it('renders correctly', () => {
    const infoTooltip = render(<InfoTooltip label={'label'} />)

    expect(infoTooltip).toMatchSnapshot()
  })
})
