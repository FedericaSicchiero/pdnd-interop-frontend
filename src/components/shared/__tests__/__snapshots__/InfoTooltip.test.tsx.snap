// Vitest Snapshot v1

exports[`Checks that InfoTooltip snapshots don't change > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <svg
        aria-hidden="true"
        aria-label="label"
        class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeSmall css-kpimro-MuiSvgIcon-root"
        data-mui-internal-clone-element="true"
        data-testid="InfoIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
        />
      </svg>
    </div>
  </body>,
  "container": <div>
    <svg
      aria-hidden="true"
      aria-label="label"
      class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeSmall css-kpimro-MuiSvgIcon-root"
      data-mui-internal-clone-element="true"
      data-testid="InfoIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
      />
    </svg>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
